### ISSUE TITLE
Unable to reproduce results

### ISSUE THREAD

**Issue by xsacha** [Herosan163/AgeEstimation#1]
Log from last two epochs

> epoch: 97, mean_loss: 11.779, variance_loss: 81.855, softmax_loss: 4.167, loss: 97.801, mae: 4.265306
> epoch: 97, test_mae: 3.533333
> [98,     0] mean_loss: 0.015, variance_loss: 0.086, softmax_loss: 0.075, loss: 0.176
> [98,     1] mean_loss: 0.074, variance_loss: 0.181, softmax_loss: 0.201, loss: 0.456
> [98,     2] mean_loss: 0.042, variance_loss: 0.137, softmax_loss: 0.155, loss: 0.334
> [98,     3] mean_loss: 0.018, variance_loss: 0.261, softmax_loss: 0.175, loss: 0.454
> [98,     4] mean_loss: 0.003, variance_loss: 0.084, softmax_loss: 0.084, loss: 0.171
> [98,     5] mean_loss: 0.032, variance_loss: 0.132, softmax_loss: 0.165, loss: 0.329
> [98,     6] mean_loss: 0.029, variance_loss: 0.270, softmax_loss: 0.068, loss: 0.367
> [98,     7] mean_loss: 0.142, variance_loss: 0.065, softmax_loss: 0.233, loss: 0.440
> [98,     8] mean_loss: 0.085, variance_loss: 0.187, softmax_loss: 0.177, loss: 0.449
> [98,     9] mean_loss: 0.002, variance_loss: 0.070, softmax_loss: 0.053, loss: 0.125
> [98,    10] mean_loss: 0.034, variance_loss: 0.254, softmax_loss: 0.226, loss: 0.514
> [98,    11] mean_loss: 0.003, variance_loss: 0.083, softmax_loss: 0.128, loss: 0.214
> [98,    12] mean_loss: 0.012, variance_loss: 0.081, softmax_loss: 0.173, loss: 0.265
> [98,    13] mean_loss: 0.007, variance_loss: 0.146, softmax_loss: 0.083, loss: 0.236
> epoch: 98, mean_loss: 11.781, variance_loss: 81.849, softmax_loss: 4.172, loss: 97.801, mae: 4.724490
> epoch: 98, test_mae: 2.800000
> [99,     0] mean_loss: 0.020, variance_loss: 0.038, softmax_loss: 0.143, loss: 0.200
> [99,     1] mean_loss: 0.017, variance_loss: 0.191, softmax_loss: 0.104, loss: 0.312
> [99,     2] mean_loss: 0.005, variance_loss: 0.138, softmax_loss: 0.065, loss: 0.208
> [99,     3] mean_loss: 0.001, variance_loss: 0.041, softmax_loss: 0.084, loss: 0.126
> [99,     4] mean_loss: 0.036, variance_loss: 0.171, softmax_loss: 0.111, loss: 0.319
> [99,     5] mean_loss: 0.003, variance_loss: 0.060, softmax_loss: 0.082, loss: 0.145
> [99,     6] mean_loss: 0.007, variance_loss: 0.121, softmax_loss: 0.140, loss: 0.268
> [99,     7] mean_loss: 0.060, variance_loss: 0.245, softmax_loss: 0.378, loss: 0.682
> [99,     8] mean_loss: 0.026, variance_loss: 0.323, softmax_loss: 0.127, loss: 0.476
> [99,     9] mean_loss: 0.001, variance_loss: 0.080, softmax_loss: 0.051, loss: 0.132
> [99,    10] mean_loss: 0.028, variance_loss: 0.239, softmax_loss: 0.110, loss: 0.377
> [99,    11] mean_loss: 0.102, variance_loss: 0.127, softmax_loss: 0.171, loss: 0.399
> [99,    12] mean_loss: 0.032, variance_loss: 0.118, softmax_loss: 0.196, loss: 0.345
> [99,    13] mean_loss: 0.021, variance_loss: 0.213, softmax_loss: 0.118, loss: 0.351
> epoch: 99, mean_loss: 11.747, variance_loss: 82.039, softmax_loss: 4.170, loss: 97.956, mae: 4.540816
> epoch: 99, test_mae: 4.000000
> best_loss_epoch: 2, best_val_loss: 96.452576, best_mae_epoch: 43, best_val_mae: 3.683673

As you can see it doesn't seem to be converging and the best loss was at epoch 2

**Comment by Herosan163** [comment]
Thank you for your report. 
Maybe, the loss at epoch 2 is the smallest, isn't it?

But it is strange that the loss of hardly fluctuates.
I will investigate this matter.

**Comment by xsacha** [comment]
Yes, the loss starts out at 96.45 at epoch 2 and never gets better than this. I don't think this is normal?
Even the MAE isn't stable. I see it hit 10.3 at epoch 72.

**Comment by Herosan163** [comment]
I checked my code, but I couldn't find the cause.
It takes time to find the cause.

**Comment by Herosan163** [comment]
I fixed this bug. The cause is that the softmax function is applied twice to the output of full connection layers in the evaluation phase. The softmax function is applied in mean-variance loss function, thus the output which the softmax function is not applied to should be input to mean-variance loss.

**Comment by xsacha** [comment]
After trying your latest version, there still seems to be no convergence.

> [97,     0] mean_loss: 0.111, variance_loss: 0.460, softmax_loss: 0.210, loss: 0.781
> [97,     1] mean_loss: 0.129, variance_loss: 0.464, softmax_loss: 0.374, loss: 0.967
> [97,     2] mean_loss: 0.560, variance_loss: 0.192, softmax_loss: 0.451, loss: 1.202
> [97,     3] mean_loss: 0.181, variance_loss: 0.572, softmax_loss: 0.324, loss: 1.077
> [97,     4] mean_loss: 0.095, variance_loss: 0.470, softmax_loss: 0.368, loss: 0.933
> [97,     5] mean_loss: 0.169, variance_loss: 0.316, softmax_loss: 0.497, loss: 0.981
> [97,     6] mean_loss: 0.164, variance_loss: 0.537, softmax_loss: 0.646, loss: 1.347
> [97,     7] mean_loss: 0.233, variance_loss: 1.738, softmax_loss: 0.439, loss: 2.410
> [97,     8] mean_loss: 0.379, variance_loss: 0.646, softmax_loss: 0.496, loss: 1.521
> [97,     9] mean_loss: 0.074, variance_loss: 0.275, softmax_loss: 0.459, loss: 0.808
> [97,    10] mean_loss: 0.114, variance_loss: 0.366, softmax_loss: 0.640, loss: 1.120
> [97,    11] mean_loss: 0.790, variance_loss: 0.399, softmax_loss: 1.445, loss: 2.634
> [97,    12] mean_loss: 0.287, variance_loss: 1.006, softmax_loss: 1.306, loss: 2.599
> [97,    13] mean_loss: 0.722, variance_loss: 1.485, softmax_loss: 1.081, loss: 3.288
> epoch: 97, mean_loss: 4.579, variance_loss: 2.414, softmax_loss: 11.313, loss: 18.305, mae: 9.795918
> epoch: 97, test_mae: 5.533333
> [98,     0] mean_loss: 1.388, variance_loss: 0.559, softmax_loss: 1.391, loss: 3.339
> [98,     1] mean_loss: 1.009, variance_loss: 0.636, softmax_loss: 1.375, loss: 3.020
> [98,     2] mean_loss: 0.408, variance_loss: 0.387, softmax_loss: 1.016, loss: 1.812
> [98,     3] mean_loss: 0.369, variance_loss: 0.838, softmax_loss: 1.217, loss: 2.425
> [98,     4] mean_loss: 0.230, variance_loss: 0.664, softmax_loss: 0.939, loss: 1.833
> [98,     5] mean_loss: 0.235, variance_loss: 2.137, softmax_loss: 1.000, loss: 3.372
> [98,     6] mean_loss: 0.170, variance_loss: 0.692, softmax_loss: 1.188, loss: 2.050
> [98,     7] mean_loss: 0.366, variance_loss: 1.222, softmax_loss: 0.984, loss: 2.572
> [98,     8] mean_loss: 0.882, variance_loss: 1.124, softmax_loss: 1.471, loss: 3.477
> [98,     9] mean_loss: 0.500, variance_loss: 0.908, softmax_loss: 2.308, loss: 3.715
> [98,    10] mean_loss: 0.401, variance_loss: 0.987, softmax_loss: 1.676, loss: 3.064
> [98,    11] mean_loss: 0.309, variance_loss: 0.439, softmax_loss: 1.648, loss: 2.396
> [98,    12] mean_loss: 0.551, variance_loss: 0.603, softmax_loss: 2.654, loss: 3.807
> [98,    13] mean_loss: 0.229, variance_loss: 0.699, softmax_loss: 1.986, loss: 2.914
> epoch: 98, mean_loss: 1.462, variance_loss: 1.087, softmax_loss: 7.008, loss: 9.557, mae: 5.061224
> epoch: 98, test_mae: 7.933333
> [99,     0] mean_loss: 0.198, variance_loss: 0.474, softmax_loss: 1.754, loss: 2.427
> [99,     1] mean_loss: 0.236, variance_loss: 0.870, softmax_loss: 2.070, loss: 3.175
> [99,     2] mean_loss: 1.169, variance_loss: 1.267, softmax_loss: 2.458, loss: 4.894
> [99,     3] mean_loss: 0.355, variance_loss: 1.196, softmax_loss: 1.688, loss: 3.239
> [99,     4] mean_loss: 0.372, variance_loss: 0.672, softmax_loss: 1.724, loss: 2.768
> [99,     5] mean_loss: 0.384, variance_loss: 1.011, softmax_loss: 1.956, loss: 3.351
> [99,     6] mean_loss: 0.378, variance_loss: 1.114, softmax_loss: 1.451, loss: 2.943
> [99,     7] mean_loss: 0.607, variance_loss: 0.715, softmax_loss: 2.247, loss: 3.569
> [99,     8] mean_loss: 0.301, variance_loss: 0.689, softmax_loss: 1.343, loss: 2.333
> [99,     9] mean_loss: 0.316, variance_loss: 0.592, softmax_loss: 1.513, loss: 2.420
> [99,    10] mean_loss: 0.330, variance_loss: 1.034, softmax_loss: 1.362, loss: 2.725
> [99,    11] mean_loss: 0.608, variance_loss: 0.690, softmax_loss: 1.754, loss: 3.052
> [99,    12] mean_loss: 0.203, variance_loss: 0.711, softmax_loss: 1.320, loss: 2.234
> [99,    13] mean_loss: 0.371, variance_loss: 0.942, softmax_loss: 1.705, loss: 3.018
> epoch: 99, mean_loss: 1.972, variance_loss: 2.338, softmax_loss: 7.897, loss: 12.207, mae: 5.897959
> epoch: 99, test_mae: 7.133333
> best_loss_epoch: 36, best_val_loss: 5.913299, best_mae_epoch: 36, best_val_mae: 3.540816

This is the last three epochs. In your graphs you show the MAE decreasing and staying stable. For me, it changes drastically at each epoch and even by epoch 97 you can see val MAE as 9.8.

One improvement: the loss seems to be related to MAE now.

**Comment by Herosan163** [comment]
It is natural that you cannot reproduce my results, because random numbers are used in the dataloader function, the dropout function and the initial parameter weights. If you reproduce results, please set these random seeds to specific numbers whenever you use them.

**Comment by xsacha** [comment]
I don't expect to get the exact numbers you get, but I expect the graph to look similar to yours.
Your graph shows MAE starting high and converging to somewhere between 2.5-5 by epoch 60 and staying there until epoch 100.
Meanwhile, mine just moves randomly up and down and never converges.
https://user-images.githubusercontent.com/53385884/62007786-ae573780-b18c-11e9-9b87-b56bc4b6a014.png

**Comment by Herosan163** [comment]
Maybe, it is due to learning rate scheduler. I updated main.py.
before: lr_scheduler.MultiStepLR(optimizer, milestones=[120], gamma=0.1)
after: lr_scheduler.MultiStepLR(optimizer, milestones=[70, 90], gamma=0.1)

**Comment by xsacha** [comment]
Thanks, graph looks a bit better now. It seems to always get best epoch around 40 though so maybe milestone should be earlier?

**Comment by Herosan163** [comment]
Yes, you may get converged results earlier.

**Comment by xsacha** [comment]
The graph looks normal now in terms of loss/MAE but I still get very different results.
Considering we have the same random seed (2019) now I'm not expecting any differences.

My training result at epoch 100:
best_loss_epoch: 46, best_val_loss: 6.134555, best_mae_epoch: 93, best_val_mae: 3.673469

The example, subject #1 that you give:
```
python3 main.py -pi /d/FGNET/images/001A43a.JPG -pm result/model_best_loss
Age: 27
```

Yours: 41

Mine is out by 15, yours is out by 2 in this example. I also tried the epoch 93 result:
```
python3 main.py -pi /d/FGNET/images/001A43a.JPG -pm result/model_best_mae
Age: 26
```

I trained with Python 3 on Linux (Ubuntu 18.04) and a Titan V using CUDA 10.1. Pytorch v1.2.0.

I think the error may be caused by:

> /usr/local/lib/python3.6/dist-packages/torch/optim/lr_scheduler.py:82: UserWarning: Detected call of `lr_scheduler.step()` before `optimizer.step()`. In PyTorch 1.1.0 and later, you should call them in the opposite order: `optimizer.step()` before `lr_scheduler.step()`.  Failure to do this will result in PyTorch skipping the first value of the learning rate schedule.See more details at https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate

So as a test I have done this:
```Diff
diff --git a/main.py b/main.py
index b998447..b409bb1 100644
--- a/main.py
+++ b/main.py
@@ -244,8 +244,8 @@ def main():
             if epoch == 10:
                 for param in model.parameters():
                     param.requires_grad = True
-            scheduler.step(epoch)
             train(train_loader, model, criterion1, criterion2, optimizer, epoch, args.result_directory)
+            scheduler.step(epoch)
             mean_loss, variance_loss, softmax_loss, loss_val, mae = evaluate(val_loader, model, criterion1, criterion2)
             mae_test = test(test_loader, model)
             print('epoch: %d, mean_loss: %.3f, variance_loss: %.3f, softmax_loss: %.3f, loss: %.3f, mae: %3f' %
```

I now get:
best_loss_epoch: 30, best_val_loss: 6.484368, best_mae_epoch: 86, best_val_mae: 3.857143
```
python3 main.py -pi /d/FGNET/images/001A43a.JPG -pm result/model_best_loss
Age: 30
```
```
python3 main.py -pi /d/FGNET/images/001A43a.JPG -pm result/model_best_mae
Age: 39
```

I'm still not quite happy with it. Any idea what I could be doing wrong?

**Comment by xsacha** [comment]
Just a note, I had to increase the initial learning rate to 0.01 to get things working better.
The readme suggested to use 0.002.

> best_loss_epoch: 46, best_val_loss: 5.397519, best_mae_epoch: 51, best_val_mae: 3.387755

> python3 main.py -pi /d/FGNET/images/001A43a.JPG -pm result/model_best_loss
> Age: 40
> python3 main.py -pi /d/FGNET/images/001A43a.JPG -pm result/model_best_mae
> Age: 40

**Comment by xsacha** [comment]
Also, some notes:
I changed the backend architecture to EfficientNet-b1 and optimiser to Adam. Then changed input size to 128 (with centecrop from 160). Cropped image helps a lot because the dataset has a lot of background in the image right now.

I get much better loss and MAE now. Especially on real data (rather than FGNET).

> epoch: 99, mean_loss: 0.688, variance_loss: 0.572, softmax_loss: 3.806, loss: 5.065, mae: 3.326531
> epoch: 99, test_mae: 2.736842
> best_loss_epoch: 45, best_val_loss: 4.765288, best_mae_epoch: 86, best_val_mae: 3.306122
